src/streams_reader.c

```diff
@@ -5,39 +5,270 @@
 #include "redisgears_memory.h"
 #include "lock_handler.h"
 #include "utils/dict.h"
+#include "execution_plan.h"
+#include "record.h"
 
 #define STREAM_REGISTRATION_INIT_SIZE 10
 Gears_list* streamsRegistration = NULL;
 
+typedef struct StreamId{
+        long first;
+        long second;
+}StreamId;
+
 typedef struct StreamReaderCtx{
     char* streamKeyName;
-    char* lastId;
-    Record** records;
+    char* consumerGroup;
+    char* streamId;
+    bool readPenging;
+    size_t batchSize;
+    Gears_list* records;
+    bool isDone;
+    RedisModuleString** batchIds;
+    StreamId lastReadId;
 }StreamReaderCtx;
 
-typedef struct StreamReaderTrigger{
-    char* streamKeyName;
-    Gears_dict* lastIds;
+typedef struct StreamReaderTriggerArgs{
+    size_t batchSize;
+    size_t durationMS;
+    char* stream;
+}StreamReaderTriggerArgs;
+
+typedef struct StreamReaderTriggerCtx{
+    size_t refCount;
+    StreamReaderTriggerArgs* args;
+    Gears_dict* singleStreamData;
     FlatExecutionPlan* fep;
-}StreamReaderTrigger;
+    ExecutionMode mode;
+    long long numTriggered;
+    long long numAborted;
+    long long numSuccess;
+    long long numFailures;
+    char* lastError;
+    pthread_t scanThread;
+    volatile bool allowStartExecution;
+}StreamReaderTriggerCtx;
+
+typedef struct SingleStreamReaderCtx{
+    size_t numTriggered;
+    bool timerIsSet;
+    RedisModuleTimerID lastTimerId;
+    StreamReaderTriggerCtx* srtctx; // weak ptr to the StreamReaderCtx
+    char* keyName;
+    StreamId lastId;
+}SingleStreamReaderCtx;
+
+static bool StreamReader_VerifyCallReply(RedisModuleCtx* ctx, RedisModuleCallReply* reply, const char* msgPrefix, const char* logLevel){
+    if (reply == NULL ||
+            RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_ERROR ||
+            RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_NULL) {
+        if(!msgPrefix){
+            // no log message requested, just return false
+            return false;
+        }
+        const char* err = "got null reply";
+        if(reply && RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_ERROR){
+            err = RedisModule_CallReplyStringPtr(reply, NULL);
+        }
+        RedisModule_Log(ctx, logLevel, "%s : %s", msgPrefix, err);
+        if(reply) RedisModule_FreeCallReply(reply);
+        return false;
+    }
+    return true;
+}
+
+static void StreamReader_RunOnEvent(SingleStreamReaderCtx* ssrctx, size_t batch, bool readPending);
+
+#define StreamIdZero (StreamId){.first = 0, .second = 0}
+
+#define StreamIdIsZero(id) (id.first == 0 && id.second == 0)
+
+static int StreamReader_StreamIdCompare(StreamId id1, StreamId id2){
+    if(id1.second > id2.second){
+        return 1;
+    }
+
+    if(id1.second < id2.second){
+        return -1;
+    }
+
+    if(id1.first > id2.first){
+        return 1;
+    }
+
+    if(id1.first < id2.first){
+        return -1;
+    }
+
+    return 0;
+}
+
+static StreamId StreamReader_ParseStreamId(const char* streamId){
```

Comment:
Maybe it's better to have an invalid streamid and to return it on failure?

Reply:
I can return NULL on error and assert at outside of the function, sounds good?